src/registers/model_specific.rs

```diff
@@ -172,4 +203,125 @@ mod x86_64 {
+        /// RPL bits 1:0 should be initialized to 11b.
+        /// - Field 2 (SYSCALL): This field is copied directly into CS.Sel. SS.Sel is set to
+        ///  this field + 8. Because SYSCALL always switches to CPL 0, the RPL bits
+        /// 33:32 should be initialized to 00b.
+        pub fn read_raw() -> (u64, u64) {
+            let msr_value = unsafe { Self::MSR.read() };
+            let sysret = msr_value.get_bits(48..63);
+            let syscall = msr_value.get_bits(32..47);
+            (sysret, syscall)
+        }
+
+        /// Read the Ring 0 and Ring 3 segment bases.
+        /// Returns
+        /// - CS Selector SYSRET
+        /// - SS Selector SYSRET
+        /// - CS Selector SYSCALL
+        /// - SS Selector SYSCALL
+        pub fn read() -> (
+            SegmentSelector,
+            SegmentSelector,
+            SegmentSelector,
+            SegmentSelector,
+        ) {
+            let raw = Self::read_raw();
+            return (
+                SegmentSelector((raw.0 + 16 - 3).try_into().unwrap()),
+                SegmentSelector((raw.0 + 8 - 3).try_into().unwrap()),
+                SegmentSelector((raw.1).try_into().unwrap()),
+                SegmentSelector((raw.1 + 8).try_into().unwrap()),
+            );
+        }
+
+        /// Write the Ring 0 and Ring 3 segment bases.
+        /// The remaining fields are ignored because they are
+        /// not valid for long mode.
+        ///
+        /// # Parameters
+        /// - sysret: The CS selector is set to this field + 16. SS.Sel is set to
+        /// this field + 8. Because SYSRET always returns to CPL 3, the
+        /// RPL bits 1:0 should be initialized to 11b.
+        /// - syscall: This field is copied directly into CS.Sel. SS.Sel is set to
+        ///  this field + 8. Because SYSCALL always switches to CPL 0, the RPL bits
+        /// 33:32 should be initialized to 00b.
+        pub fn write_raw(sysret: u64, syscall: u64) {
+            let mut msr_value = 0u64;
+            msr_value.set_bits(48..63, sysret);
+            msr_value.set_bits(32..47, syscall);
+            unsafe {
+                Self::MSR.write(msr_value);
+            }
+        }
+
+        /// Write the Ring 0 and Ring 3 segment bases.
+        /// The remaining fields are ignored because they are
+        /// not valid for long mode.
+        pub fn write(
+            cs_sysret: SegmentSelector,
+            ss_sysret: SegmentSelector,
+            cs_syscall: SegmentSelector,
+            ss_syscall: SegmentSelector,
+        ) -> Result<(), &'static str> {
+            if cs_sysret.0 - 16 != ss_sysret.0 - 8 {
+                return Err("Sysret CS and SS is not offset by 8.");
+            }
+
+            if cs_syscall.0 != ss_syscall.0 - 8 {
+                return Err("Syscall CS and SS is not offset by 8.");
+            }
+
+            Self::write_raw(((ss_sysret.0 - 8) | 0b11).into(), cs_syscall.0.into());
```

Comment:
I'm not sure if automatically setting the RPL to `0b11` is the right solution. The alternative would be to require the caller to properly set these bits. What do you think is the better solution?

Reply:
The amd64 documentation says it is required but will not be tested. That is why i implicitly left it like this. This also provides direct one to one mapping to the user segment functions from gdt. Also the user can use write_raw to test it otherwise.

Reply:
> This also provides direct one to one mapping to the user segment functions from gdt.

The user segment functions only return a GDT descriptor, not a segment selector. You mean the `add_entry` function? If yes, there is currently an open issue about whether we should hardcode the requested privilege level in that function: https://github.com/rust-osdev/x86_64/issues/120